core/types/calls.cc

```diff
@@ -879,6 +879,19 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, const DispatchArgs &arg
             ? guessOverload(gs, symbol, mayBeOverloaded, args.numPosArgs, args.args, targs, args.block != nullptr)
             : mayBeOverloaded;
 
+    if (method.data(gs)->flags.isPrivate && !args.isPrivateOk) {
+        if (auto e = gs.beginError(errLoc, core::errors::Infer::PrivateMethod)) {
+            if (args.fullType.type != args.thisType) {
+                e.setHeader("Non-private call to private method `{}` on `{}` component of `{}`",
+                            method.data(gs)->name.show(gs), args.thisType.show(gs), args.fullType.type.show(gs));
+            } else {
+                e.setHeader("Non-private call to private method `{}` on `{}`", method.data(gs)->name.show(gs),
+                            args.thisType.show(gs));
+            }
+            e.addErrorLine(method.data(gs)->loc(), "Defined in `{}` here", method.data(gs)->owner.show(gs));
+        }
+    }
```

Comment:
Moving this snippet from environment.cc to here means that we can check privateness per `dispatchCall`, which means that if we ever make a recursive call to `dispatchCall` in some intrinsic, we have more control over whether calls to private methods are allowed.

This is important because

```ruby
class A
  def initialize; end
end

a = A.new # ok
a.initialize # not okay
```

And previously, the `Class_new` intrinsic would record the `A#initialize` symbol as "the thing that was dispatched to" in the DispatchResult, which meant that Sorbet couldn't easily tell in environment.cc whether the call to `initialize` happened at a location where private was ok (`A.new`) or not ok (`a.initialize`).

While I did this refactor, I audited all recursive calls to `dispatchCall` to see whether we should forward along the `isPrivateOk` from the call site to the recursive call. The Ruby VM allows basically all the implicit conversion methods to be called even if they're private, so you see some of those flip in this change (I didn't add tests for that, but I could if we think it's valuable).